client/ClientDecisionService/DecisionServiceLocal.cs

```diff
@@ -0,0 +1,106 @@
+﻿using Microsoft.Research.MultiWorldTesting.Contract;
+using Microsoft.Research.MultiWorldTesting.JoinUploader;
+using Microsoft.Research.MultiWorldTesting.ExploreLibrary;
+using VW;
+using VW.Labels;
+using VW.Serializer;
+using System;
+using System.IO;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.Caching;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Timers;
+using System.Diagnostics;
+
+namespace Microsoft.Research.MultiWorldTesting.ClientLibrary
+{
+    public class DecisionServiceLocal<TContext> : DecisionServiceClient<TContext>
+    {
+        private VowpalWabbit<TContext> vw;
+        // This serves as the base class's recorder/logger as well, but we keep a reference around
+        // becauses it exposes additional APIs that aren't part of those interfaces (yet)
+        private InMemoryLogger<TContext, int[]> log;
+
+        public int ModelUpdateInterval;
+        private int sinceLastUpdate = 0;
+
+        // A snapshot of the current VW model
+        public byte[] Model
+        {
+            get
+            {
+                MemoryStream currModel = new MemoryStream();
+                vw.Native.SaveModel(currModel);
+                return currModel.ToArray();
+            }
+        }
+
+        public DecisionServiceLocal(string vwArgs, int modelUpdateInterval, TimeSpan expUnit)
+            : base(
+            new DecisionServiceConfiguration("")
+            {
+                OfflineMode = true,
+                OfflineApplicationID = Guid.NewGuid().ToString(),
+                DevelopmentMode = false
+            },
+            new ApplicationClientMetadata
+            {
+                TrainArguments = vwArgs,
+                InitialExplorationEpsilon = 1f
+            },
+            new VWExplorer<TContext>(null, JsonTypeInspector.Default, false))
+        {
+            this.log = new InMemoryLogger<TContext, int[]>(expUnit);
+            this.Recorder = log;
+            this.vw = new VowpalWabbit<TContext>(
+                new VowpalWabbitSettings(vwArgs)
+                {
+                    TypeInspector = JsonTypeInspector.Default,
+                    EnableStringExampleGeneration = true,
+                    EnableStringFloatCompact = true
+                }
+                );
+            this.ModelUpdateInterval = modelUpdateInterval;
+        }
+
+        /// <summary>
+        /// Report a simple float reward for the experimental unit identified by the given unique key.
+        /// </summary>
+        /// <param name="reward">The simple float reward.</param>
+        /// <param name="uniqueKey">The unique key of the experimental unit.</param>
+        new public void ReportReward(float reward, string uniqueKey)

```

Comment:
I switched to using virtual/override. In general I agree with your suggestion of using a base class that the local and remote versions derive from. However I think this will create substantial churn in the DecisionServiceClient code for something we are not quite using yet outside of simulation (the localmode version). My suggestion is that we keep the original code as unmodified as possible until we have a proper use case for this, and then consider the refactor -- what do you think?